contrib/lkt/language/parser.py

```diff
@@ -1192,109 +1192,69 @@ def lexer():
-            lexer_annot.then(
-                lambda _: No(T.SemanticResult.array),
-                default_val=Self.error(
-                    S('missing mandatory "@with_lexer" annotation')
-                ).singleton
-            )
-        )
+        lexer_annot_params = Var(lexer_annot._.params.params)
+        lexer_arg = Var(lexer_annot_params.at(0))
 
-        # Check has expected arguments (exactly one, positional)
-        params = Var(lexer_annot._.params.params)
-        lexer_arg = Var(
-            If(
-                Not(params.is_null)
-                & (params.children.length == 1)
-                & params.at(0).name.is_null,
-                params.at(0).value,
-                No(T.Expr)
-            )
-        )
-        invalid_args = Var(
-            If(
-                Not(params.is_null) & lexer_arg.is_null,
-                lexer_annot.error(
-                    S("exactly 1 positional argument expected")
-                ).singleton,
-                No(T.SemanticResult.array),
-            )
-        )
+        # Lexer decl
+        lexer_decl = Var(lexer_arg._.value.cast(T.RefId)._.as_entity
+                         .referenced_decl.result_ref.cast(T.LexerDecl))
 
-        # Check that it is a valid reference to a lexer
-        lexer_ref = Var(lexer_arg.cast(T.RefId))
-        invalid_lexer_ref = Var(
-            If(
-                Not(lexer_arg.is_null) & lexer_ref.is_null,
-                lexer_arg.error(
-                    S("reference to a lexer expected")
-                ).singleton,
-                No(T.SemanticResult.array),
-            )
-        )
-        refd_lexer = Var(lexer_ref._.as_entity.referenced_decl)
-        lexer_not_found = Var(
-            If(
-                refd_lexer.has_error,
-                refd_lexer.singleton,
-                No(T.SemanticResult.array),
-            )
-        )
-        lexer_decl = Var(
-            If(
-                refd_lexer.has_error,
-                No(T.LexerDecl.entity),
-                refd_lexer.result_ref.cast(T.LexerDecl)
-            )
-        )
-        invalid_lexer_decl = Var(
-            If(
-                Not(lexer_ref.is_null) & lexer_decl.is_null,
-                lexer_ref.error(S("reference to a lexer expected")).singleton,
-                No(T.SemanticResult.array)
+        return (
+            # All grammars must have an associated lexer. First check that the
+            # @with_lexer annotation is present.
+            lexer_annot.is_null.then(
+                lambda _:
+                Self.error(S('missing mandatory "@with_lexer" annotation'))
+                .singleton
             )
-        )
 
-        # Now check the grammar rules themselves
-        rules = Var(Entity.rules.filter(
-            lambda d: d.decl.is_a(T.GrammarRuleDecl)
-        ))
+            # Then check that the associated arguments are correct
+            .concat(((lexer_annot_params.children.length != 1)
```

Comment:
This crashes when `lexer_annot_params` is null (previous code checked `Not(params.is_null)`)